File: backend/app/models/contracts.py
from typing import Literal

from pydantic import BaseModel, Field

Crop = Literal["wheat", "mustard", "chickpea", "potato", "cotton", "rice"]
Channel = Literal["whatsapp", "sms", "ivr", "field_rep", "retailer"]
Objective = Literal["awareness", "lead_generation", "retailer_sellthrough", "field_visit"]
Language = Literal["Hindi", "Punjabi", "Marathi", "Gujarati", "Kannada", "Bengali", "English"]
DeviceType = Literal["smartphone", "keypad", "unknown"]
RiskLevel = Literal["low", "medium", "high"]
StockStatus = Literal["healthy", "watch", "low", "out_of_stock"]
SourceMode = Literal["mock", "rules", "ml", "hybrid"]
ContentFormat = Literal["whatsapp", "sms", "ivr", "rep_script", "visual_concept"]
ApprovalState = Literal["pending_review", "approved", "rejected"]
Role = Literal["campaign_manager", "territory_manager", "field_rep", "retailer_support"]


class Geography(BaseModel):
    state: str
    district: str
    tehsil: str | None = None
    territory_id: str | None = None


class Audience(BaseModel):
    languages: list[Language]
    device_types: list[DeviceType]
    max_target_count: int | None = Field(default=None, ge=1)


class Constraints(BaseModel):
    low_bandwidth: bool = True
    human_review_required: bool = True
    min_stock_cover_days: int = Field(default=10, ge=0)


class CampaignContextRequest(BaseModel):
    scenario_id: str | None = None
    crop: Crop
    product: str | None = None
    objective: Objective
    week_start_date: str
    geography: Geography
    audience: Audience
    channel_preferences: list[Channel]
    constraints: Constraints


class ApiEnvelope(BaseModel):
    schema_version: str = "syngenta-copilot.v1"
    request_id: str
    generated_at: str
    source_mode: SourceMode
    warnings: list[str] = []


class CropStage(BaseModel):
    stage: str
    days_to_stage: int
    confidence: float = Field(ge=0, le=1)


class GrowerSummary(BaseModel):
    estimated_growers: int = Field(ge=0)
    smartphone_share: float = Field(ge=0, le=1)
    keypad_share: float = Field(ge=0, le=1)
    primary_language: str


class WeatherInsight(BaseModel):
    risk_type: str
    risk_level: RiskLevel
    summary: str
    confidence: float = Field(ge=0, le=1)


class InventoryAlert(BaseModel):
    product: str
    stock_status: StockStatus
    stock_cover_days: int = Field(ge=0)
    affected_retailers: int = Field(ge=0)


class CampaignContextResponse(ApiEnvelope):
    context_id: str
    crop_stage: CropStage
    grower_summary: GrowerSummary
    weather_insights: list[WeatherInsight]
    inventory_alerts: list[InventoryAlert]


class Receptivity(BaseModel):
    open_probability: float = Field(ge=0, le=1)
    click_probability: float = Field(ge=0, le=1)
    confidence: float = Field(ge=0, le=1)


class ChannelStrategy(BaseModel):
    channel: Channel
    rank: int = Field(ge=1)
    reason: str


class Timing(BaseModel):
    recommended_send_date: str
    send_window: str
    urgency: RiskLevel


class ExpectedImpact(BaseModel):
    baseline_click_rate: float = Field(ge=0, le=1)
    expected_click_rate: float = Field(ge=0, le=1)
    expected_leads: int = Field(ge=0)


class Recommendation(BaseModel):
    recommendation_id: str
    priority_score: int = Field(ge=0, le=100)
    segment_label: str
    target_count: int = Field(ge=0)
    crop: Crop
    product: str
    channel_strategy: list[ChannelStrategy]
    timing: Timing
    receptivity: Receptivity
    expected_impact: ExpectedImpact
    reason_codes: list[str]
    human_review_flags: list[str]
    blocked: bool = False


class RecommendationResponse(ApiEnvelope):
    plan_id: str
    context_id: str
    recommendations: list[Recommendation]


class ContentGenerationRequest(BaseModel):
    plan_id: str
    recommendation_id: str
    languages: list[Language] = ["Hindi"]
    formats: list[ContentFormat]
    tone: str = "trusted_advisory"


class ContentVariant(BaseModel):
    content_id: str
    format: ContentFormat
    language: Language
    text: str
    cta: str | None = None
    estimated_read_time_sec: int | None = Field(default=None, ge=0)
    approval_state: ApprovalState = "pending_review"
    safety_flags: list[str] = []
    generation_source: str | None = None
    fallback_reason: str | None = None


class ContentGenerationResponse(ApiEnvelope):
    content_batch_id: str
    plan_id: str
    recommendation_id: str
    variants: list[ContentVariant]


class ContentApprovalRequest(BaseModel):
    content_id: str
    approval_state: Literal["approved", "rejected"]
    reviewer: str | None = None


class ContentApprovalResponse(ApiEnvelope):
    content_id: str
    content_batch_id: str
    plan_id: str
    recommendation_id: str
    approval_state: ApprovalState
    reviewer: str | None = None
    approved_at: str | None = None
    field_actions_unlocked: bool


class FieldAction(BaseModel):
    action_id: str
    rep_id: str
    territory_id: str
    priority: RiskLevel
    due_date: str
    action_type: str
    summary: str
    retailer_ids: list[str]
    recommended_script_id: str
    success_metric: str


class FieldActionsResponse(ApiEnvelope):
    plan_id: str
    actions: list[FieldAction]


class AnalyticsSummaryResponse(ApiEnvelope):
    plan_id: str
    kpis: dict
    charts: dict


class ScenarioResponse(ApiEnvelope):
    scenarios: list["Scenario"]


class Scenario(BaseModel):
    scenario_id: str
    name: str
    crop: Crop
    geography: Geography
    description: str
    risk_level: RiskLevel
    stock_status: StockStatus


class ExportRequest(BaseModel):
    plan_id: str = "PLAN_001"
    export_type: Literal["csv", "whatsapp_pack", "rep_brief"] = "csv"


class ExportResponse(ApiEnvelope):
    export_id: str
    plan_id: str
    export_type: Literal["csv", "whatsapp_pack", "rep_brief"]
    formats: list[str]
    download_url: str


# ---------------------------------------------------------------------------
# Workflow Orchestration Contracts
# ---------------------------------------------------------------------------

class WorkflowStartRequest(BaseModel):
    """Request to bootstrap a full workflow chain."""
    scenario_id: str | None = None
    crop: Crop = "wheat"
    product: str = "Tilt 250 EC"
    objective: Objective = "lead_generation"
    week_start_date: str = "2026-02-16"
    geography: Geography = Geography(state="Uttar Pradesh", district="Kanpur Nagar")
    audience: Audience = Audience(languages=["Hindi"], device_types=["smartphone"])
    channel_preferences: list[Channel] = ["whatsapp", "sms", "field_rep"]
    constraints: Constraints = Constraints()
    role: Role = "campaign_manager"


class WorkflowEvent(BaseModel):
    """A single workflow lifecycle event."""
    event_id: str
    workflow_id: str
    event_type: str
    role: str | None = None
    timestamp: str
    severity: RiskLevel = "low"
    description: str
    territory: str | None = None
    metadata: dict = {}


class NextBestAction(BaseModel):
    """Recommended next action based on workflow state."""
    action: str
    reason: str
    assigned_role: str
    priority: RiskLevel
    target_entity_id: str | None = None


class KpiItem(BaseModel):
    """A single KPI metric."""
    label: str
    value: str
    trend: str
    metadata: str | None = None
    tone: str = "ai"
    source: str = "computed"


class KpiData(BaseModel):
    """Role-aware KPI collection."""
    role: str
    kpis: list[KpiItem]


class OperationalEvent(BaseModel):
    """An operational intelligence event for the live feed."""
    event_id: str
    text: str
    time: str
    event_type: str = "info"
    role: str | None = None
    severity: RiskLevel = "low"


class SystemHealth(BaseModel):
    """Backend system health snapshot."""
    gemini: str = "unknown"
    supabase: str = "unknown"
    cache: str = "unknown"
    data_mode: str = "unknown"
    active_workflows: int = 0
    last_generation_source: str = "unknown"


class WorkflowState(ApiEnvelope):
    """Complete workflow state returned by the orchestrator."""
    workflow_id: str
    plan_id: str
    context_id: str
    status: str = "active"
    context: CampaignContextResponse | None = None
    recommendations: list[Recommendation] = []
    content_variants: list[ContentVariant] = []
    events: list[WorkflowEvent] = []
    kpis: KpiData | None = None
    alerts: list[OperationalEvent] = []
    next_action: NextBestAction | None = None
    system_health: SystemHealth | None = None

File: frontend/src/types/contracts.ts
export type Crop = "wheat" | "mustard" | "chickpea" | "potato" | "cotton" | "rice";
export type Channel = "whatsapp" | "sms" | "ivr" | "field_rep" | "retailer";
export type Objective = "awareness" | "lead_generation" | "retailer_sellthrough" | "field_visit";
export type Language = "Hindi" | "Punjabi" | "Marathi" | "Gujarati" | "Kannada" | "Bengali" | "English";
export type DeviceType = "smartphone" | "keypad" | "unknown";
export type RiskLevel = "low" | "medium" | "high";
export type StockStatus = "healthy" | "watch" | "low" | "out_of_stock";
export type SourceMode = "mock" | "rules" | "ml" | "hybrid";
export type ContentFormat = "whatsapp" | "sms" | "ivr" | "rep_script" | "visual_concept";
export type ApprovalState = "pending_review" | "approved" | "rejected";
export type Role = "campaign_manager" | "territory_manager" | "field_rep" | "retailer_support";

export interface ApiEnvelope {
  schema_version: "syngenta-copilot.v1";
  request_id: string;
  generated_at: string;
  source_mode?: string;
  warnings: string[];
  model_version?: string;
  trained_on?: string;
  feature_version?: string;
  data_last_updated?: string;
  inventory_snapshot?: string;
  model_last_trained?: string;
  response_time_ms?: number;
}

export interface Geography {
  state: string;
  district: string;
  tehsil?: string;
  territory_id?: string;
}

export interface CampaignContextRequest {
  scenario_id?: string;
  crop: Crop;
  product?: string;
  objective: Objective;
  week_start_date: string;
  geography: Geography;
  audience: {
    languages: Language[];
    device_types: DeviceType[];
    max_target_count?: number;
  };
  channel_preferences: Channel[];
  constraints: {
    low_bandwidth: boolean;
    human_review_required: boolean;
    min_stock_cover_days: number;
  };
}

export interface CampaignContextResponse extends ApiEnvelope {
  context_id: string;
  crop_stage: {
    stage: string;
    days_to_stage: number;
    confidence: number;
  };
  grower_summary: {
    estimated_growers: number;
    smartphone_share: number;
    keypad_share: number;
    primary_language: Language;
  };
  weather_insights: WeatherInsight[];
  inventory_alerts: InventoryAlert[];
}

export interface WeatherInsight {
  risk_type: string;
  risk_level: RiskLevel;
  summary: string;
  confidence: number;
}

export interface InventoryAlert {
  product: string;
  stock_status: StockStatus;
  stock_cover_days: number;
  affected_retailers: number;
}

export interface ExecutiveSummary {
  total_target_growers: number;
  predicted_open_rate: number;
  predicted_click_rate: number;
  expected_leads: number;
  stock_ready_retailers: number;
  summary_text: string;
}

export interface RecommendationResponse extends ApiEnvelope {
  plan_id: string;
  context_id: string;
  recommendations: Recommendation[];
  executive_summary?: ExecutiveSummary;
}

export interface DataQualityWarning {
  message: string;
  severity: "low" | "medium" | "high";
}

export interface Recommendation {
  recommendation_id: string;
  plan_id?: string;
  priority_score: number;
  segment_label: string;
  target_count: number;
  crop: Crop;
  product: string;
  channel_strategy: { channel: Channel; rank: number; reason: string }[];
  timing: {
    recommended_send_date: string | null;
    send_window: string | null;
    urgency: RiskLevel | "blocked";
  };
  receptivity: {
    open_probability: number | null;
    click_probability: number | null;
    confidence: string | null;
    confidence_label?: string;
  };
  expected_impact: {
    baseline_click_rate: number;
    expected_click_rate: number;
    expected_leads: number;
  };
  reason_codes: string[];
  human_review_flags: string[];
  blocked: boolean;
  blocked_reasons?: string[];
  actionability_status?: "Ready to Execute" | "Needs Human Review" | "Blocked";
  data_quality_warnings?: DataQualityWarning[];
  operational_readiness_score?: number;
  recommendation_priority_rank?: number;
}

export interface ContentGenerationResponse extends ApiEnvelope {
  content_batch_id: string;
  plan_id: string;
  recommendation_id: string;
  variants: ContentVariant[];
}

export interface ContentVariant {
  content_id: string;
  format: ContentFormat;
  language: Language;
  text: string;
  cta?: string;
  estimated_read_time_sec?: number;
  approval_state: ApprovalState;
  safety_flags: string[];
}

export interface ContentApprovalResponse extends ApiEnvelope {
  content_id: string;
  content_batch_id: string;
  plan_id: string;
  recommendation_id: string;
  approval_state: ApprovalState;
  reviewer?: string;
  approved_at?: string;
  field_actions_unlocked: boolean;
}

export interface FieldActionsResponse extends ApiEnvelope {
  plan_id: string;
  actions: FieldAction[];
}

export interface FieldAction {
  action_id: string;
  rep_id: string;
  territory_id: string;
  priority: RiskLevel;
  due_date: string;
  action_type: string;
  summary: string;
  retailer_ids: string[];
  recommended_script_id: string;
  success_metric: string;
}

export interface AnalyticsSummaryResponse extends ApiEnvelope {
  plan_id: string;
  kpis: {
    target_growers: number;
    predicted_open_rate: number;
    predicted_click_rate: number;
    expected_leads: number;
    stock_ready_retailers: number;
    field_actions: number;
  };
  charts: {
    channel_mix: { channel: Channel; share: number }[];
    weekly_funnel: { week: string; baseline: number; recommended: number }[];
    engagement_funnel?: { label: string; baseline: number; recommended: number }[];
  };
}

export interface ScenarioResponse extends ApiEnvelope {
  scenarios: Scenario[];
}

export interface Scenario {
  scenario_id: string;
  name: string;
  crop: Crop;
  geography: Geography;
  description: string;
  risk_level: RiskLevel;
  stock_status: StockStatus;
}

export interface ExportResponse extends ApiEnvelope {
  export_id: string;
  plan_id: string;
  export_type: "csv" | "whatsapp_pack" | "rep_brief";
  formats: string[];
  download_url: string;
}
